File: aisd/aisd/src/main/java/org/example/structure/hashtable/HashTableDoubleHash.java
package org.example.structure.hashtable;

import org.example.model.Person;

// Хэш таблица с двойным хэшированием
public class HashTableDoubleHash {
    private final int size; // Размер хэш-таблицы
    private final Person[] data; // Массив элементов хэш-таблицы

    public HashTableDoubleHash(int size) {
        this.size = size;
        this.data = new Person[size]; // Инициализация массива элементов
    }

    public void print() {
        int i = 1;
        System.out.println("======================");
        for (Person element : data) {
            if (element == null || element.isDeleted()) {
                System.out.println("Элемент " + i + " - null"); // Вывод информации о пустых или удаленных элементах
            } else {
                System.out.println("Элемент " + i + " - " + element.getName() + " " + element.getAge()); // Вывод информации о каждом элементе
            }
            i++;
        }
        System.out.println("======================");
    }

    public void insert(Person personModel) {
        int index = getIndex(personModel); // Получение индекса для вставки элемента

        if (data[index] == null || data[index].isDeleted()) {
            data[index] = personModel; // Вставка элемента, если ячейка пуста или содержит удаленный элемент
            return;
        }

        for (int i = 0; i < size; i++) {
            index = helperHash(personModel, i); // Вычисление нового индекса с помощью двойного хэширования
            if (data[index] == null || data[index].isDeleted()) {
                data[index] = personModel; // Вставка элемента, если ячейка пуста или содержит удаленный элемент
                return;
            }
        }

        System.out.println("Таблица переполнена"); // Вывод сообщения об ошибке, если таблица переполнена
    }

    public void delete(Person personModel) {
        for (int i = 0; i < size; i++) {
            int index = helperHash(personModel, i); // Получение индекса элемента для удаления
            if (data[index] != null && data[index].getAge() == personModel.getAge()
                    && data[index].getName().equals(personModel.getName())
                    && data[index].isDeleted() == personModel.isDeleted()) {
                data[index].setDeleted(true); // Установка флага удаления для найденного элемента
            }
        }
    }

    public boolean find(Person personModel) {
        for (int i = 0; i < size; i++) {
            int index = helperHash(personModel, i); // Получение индекса элемента для поиска
            if (data[index] == null) {
                return false; // Возвращение false, если ячейка пуста
            }
            if (data[index].getAge() == personModel.getAge()
                    && data[index].getName().equals(personModel.getName())
                    && data[index].isDeleted() == personModel.isDeleted()) {
                return true; // Возвращение true, если элемент найден
            }
        }
        return false; // Возвращение false, если элемент не найден
    }

    private int getIndex(Person personModel) {
        // Вычисление хэша для определения индекса элемента в массиве
        return (personModel.getAge() * personModel.getAge() +
                personModel.getName().length() * personModel.getName().length() +
                personModel.getAge() * personModel.getName().length()) % size;
    }

    private int helperHash(Person personModel, int step) {
        // Вычисление второго хэша с помощью двойного хэширования
        return (getIndex(personModel) + step * personModel.getAge() * personModel.getName().length()) % size;
    }
}

File: misc/ASD/ASD/HashTableDoubleHash.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ASD;
public class HashTableDoubleHash
{
    private readonly int _size;
    private readonly PersonModel?[] _data;

    public HashTableDoubleHash(int size)
    {
        _size = size;
        _data = new PersonModel?[_size];
    }

    public void Print()
    {
        int i = 1;
        foreach (var element in _data)
        {
            if (element == null || element.IsDeleted)
            {
                Console.WriteLine($"Элемент {i} - null");
            }
            else
            {
                Console.WriteLine($"Элемент {i} - {element.Name} {element.Age}");
            }
            i++;
        }
    }

    public void Insert(PersonModel personModel)
    {
        int index = GetIndex(personModel);

        if (_data[index] == null || _data[index].IsDeleted)
        {
            _data[index] = personModel;
            return;
        }

        for (int i = 0; i < _size; i++)
        {
            index = HelperHash(personModel, i);
            if (_data[index] == null || _data[index].IsDeleted)
            {
                _data[index] = personModel;
                return;
            }
        }


        Console.WriteLine("Таблица переполнена");
    }

    public void Delete(PersonModel personModel)
    {
        for (int i = 0; i < _size; i++)
        {
            int index = HelperHash(personModel, i);
            if (_data[index].Age == personModel.Age && _data[index].Name == personModel.Name && _data[index].IsDeleted == personModel.IsDeleted)
            {
                _data[index].IsDeleted = true;
            }
        }
    }

    public bool Find(PersonModel personModel)
    {


        for (int i = 0; i < _size; i++)
        {
            int index = HelperHash(personModel, i);
            if (_data[index] == null)
            {
                return false;
            }
            if (_data[index].Age == personModel.Age && _data[index].Name == personModel.Name && _data[index].IsDeleted == personModel.IsDeleted)
            {
                return true;
            }
        }

        return false;
    }

    private int GetIndex(PersonModel personModel)
        => (personModel.Age * personModel.Age + personModel.Name.Length * personModel.Name.Length + personModel.Age * personModel.Name.Length) % _size;

    private int HelperHash(PersonModel personModel, int step)
        => (GetIndex(personModel) + step * personModel.Age * personModel.Name.Length) % _size;
}
